File: src/models/beat_grid.rs
use super::time_signature::TimeSignature;

/// Represents the underlying rhythmic framework
#[derive(Debug, Clone)]
pub struct BeatGrid {
    /// Beats per measure and note value
    pub time_signature: TimeSignature,
    /// Smallest rhythmic unit (16 = sixteenth notes)
    pub subdivision: u8,
    /// Number of measures in grid
    pub num_measures: u8,
    /// Probability weights for each grid position
    pub position_weights: Vec<f32>,
}

impl BeatGrid {
    /// Create a new beat grid
    pub fn new(
        time_signature: TimeSignature,
        subdivision: u8,
        num_measures: u8,
        position_weights: Vec<f32>,
    ) -> Self {
        Self {
            time_signature,
            subdivision,
            num_measures,
            position_weights,
        }
    }

    /// Total number of grid positions
    pub fn total_positions(&self) -> usize {
        (self.subdivision as usize)
            * (self.time_signature.numerator as usize)
            * (self.num_measures as usize)
            / 4 // subdivision is per quarter note
    }

    /// Get indices of on-beat positions (0, 4, 8, 12 in 4/4 sixteenths)
    pub fn beat_positions(&self) -> Vec<usize> {
        let positions_per_beat = self.subdivision as usize / 4;
        (0..self.time_signature.numerator as usize)
            .map(|beat| beat * positions_per_beat)
            .collect()
    }

    /// Get metrical strength of a position (1.0 = downbeat, 0.0 = weakest)
    pub fn position_strength(&self, idx: usize) -> f32 {
        let positions_per_beat = self.subdivision as usize / 4;

        // Position 0 (downbeat) is strongest
        if idx == 0 {
            return 1.0;
        }

        // On-beat positions (every quarter note)
        if idx % positions_per_beat == 0 {
            // Beat 3 is stronger than beats 2 and 4
            let beat_num = idx / positions_per_beat;
            if beat_num == 2 {
                return 0.7; // Beat 3 (0-indexed as 2)
            }
            return 0.4; // Beats 2 and 4
        }

        // Off-beat positions
        0.2
    }

    /// Duration of one grid position at given tempo
    pub fn seconds_per_position(&self, tempo_bpm: u16) -> f64 {
        // Quarter note duration in seconds
        let quarter_note_seconds = 60.0 / tempo_bpm as f64;

        // Each subdivision unit relative to quarter note
        let subdivisions_per_quarter = self.subdivision as f64 / 4.0;

        quarter_note_seconds / subdivisions_per_quarter
    }
}

File: web/src/models/BeatGrid.ts
/**
 * BeatGrid Model
 * Represents the underlying rhythmic framework for pattern generation
 */

import { BeatGrid as IBeatGrid, TimeSignature } from './types';

export class BeatGrid implements IBeatGrid {
  public readonly timeSignature: TimeSignature;
  public readonly subdivision: number;
  public readonly numMeasures: number;

  constructor(timeSignature: TimeSignature, subdivision: number = 16, numMeasures: number = 1) {
    this.timeSignature = { ...timeSignature };
    this.subdivision = subdivision;
    this.numMeasures = numMeasures;
  }

  /**
   * Total number of grid positions
   */
  totalPositions(): number {
    // subdivision is relative to quarter notes (16 = sixteenth notes)
    // For time signatures with different denominators, we need to adjust
    const sixteenthsPerQuarter = this.subdivision / 4;
    const quartersPerMeasure =
      (this.timeSignature.numerator * 4) / this.timeSignature.denominator;
    return sixteenthsPerQuarter * quartersPerMeasure * this.numMeasures;
  }

  /**
   * Get number of compound beats per measure (for compound meters like 6/8)
   */
  private getCompoundBeats(): number {
    const { numerator, denominator } = this.timeSignature;

    // 6/8 is compound duple (2 beats)
    if (numerator === 6 && denominator === 8) {
      return 2;
    }

    // For simple meters, use the numerator as-is
    return numerator;
  }

  /**
   * Get indices of on-beat positions (0, 4, 8, 12 in 4/4 sixteenths)
   */
  beatPositions(): number[] {
    const totalPos = this.totalPositions();
    const compoundBeats = this.getCompoundBeats();
    const positionsPerBeat = totalPos / compoundBeats;

    const positions: number[] = [];
    for (let beat = 0; beat < compoundBeats; beat++) {
      positions.push(Math.round(beat * positionsPerBeat));
    }
    return positions;
  }

  /**
   * Get metrical strength of a position (1.0 = downbeat, 0.0 = weakest)
   * Uses time-signature-specific metrical hierarchy
   */
  positionStrength(idx: number): number {
    const totalPos = this.totalPositions();
    const compoundBeats = this.getCompoundBeats();
    const positionsPerBeat = totalPos / compoundBeats;

    // Position 0 (downbeat) is always strongest
    if (idx === 0) {
      return 1.0;
    }

    // Check if this is an on-beat position (using compound beat grouping)
    if (idx % positionsPerBeat === 0) {
      const beatNum = Math.floor(idx / positionsPerBeat);
      return this.beatStrength(beatNum);
    }

    // Off-beat positions are weakest
    return 0.2;
  }

  /**
   * Get the metrical strength of a specific beat number based on time signature
   */
  private beatStrength(beatNum: number): number {
    const { numerator, denominator } = this.timeSignature;

    switch (`${numerator}/${denominator}`) {
      // 4/4 time: strong-weak-medium-weak pattern
      case '4/4':
        switch (beatNum) {
          case 0:
            return 1.0; // Downbeat
          case 2:
            return 0.7; // Beat 3 is secondary strong
          default:
            return 0.4; // Beats 2 and 4 are weak
        }

      // 3/4 time: strong-weak-weak pattern
      case '3/4':
        return beatNum === 0 ? 1.0 : 0.4;

      // 6/8 time: compound duple (2 compound beats)
      case '6/8':
        switch (beatNum) {
          case 0:
            return 1.0; // Primary downbeat (beat 1)
          case 1:
            return 0.6; // Secondary accent (beat 2)
          default:
            return 0.3; // Should not reach here in 6/8
        }

      // 2/4 time: strong-weak pattern
      case '2/4':
        return beatNum === 0 ? 1.0 : 0.4;

      // 5/4 time: strong-weak-medium-weak-weak (3+2 or 2+3 grouping)
      case '5/4':
        switch (beatNum) {
          case 0:
            return 1.0; // Downbeat
          case 2:
            return 0.6; // Secondary accent at beat 3
          default:
            return 0.3; // Other beats weak
        }

      // Default pattern for unhandled time signatures
      default:
        if (beatNum === 0) return 1.0;
        if (beatNum === Math.floor(numerator / 2)) return 0.6;
        return 0.4;
    }
  }

  /**
   * Duration of one grid position at given tempo
   */
  secondsPerPosition(tempoBpm: number): number {
    // Quarter note duration in seconds
    const quarterNoteSeconds = 60.0 / tempoBpm;

    // Each subdivision unit relative to quarter note
    const subdivisionsPerQuarter = this.subdivision / 4.0;

    return quarterNoteSeconds / subdivisionsPerQuarter;
  }
}
